### PR TITLE
[SR-4231] Add diagnostic (& fix-it) for mixed syntax availability attribute

### PR DESCRIPTION
**Pull request by d-ronnqvist**
This adds a specific diagnostic when both shorthand and non-shorthand specifications are mixed in the same availability attribute. For example: `@available(OSX 10.0, deprecated: 10.12)`.

Current error:
```
@available(OSX 10.0, deprecated: 10.12)
//                             ^
// error: expected version number
// error: expected declaration
```

New error:
```
@available(OSX 10.0, deprecated: 10.12)
//                   ^
// error: 'deprecated' can't be combined with shorthand specification 'OSX 10.0'
// error: expected declaration
```

If (and only if) there is only a single platform version constraint, that is followed by ‘deprecated’, ‘renamed’, etc. but not ‘introduced’, then it also offers a fix-it to insert `, introduced:` in the platform version constraint, turning:
```
@available(OSX 10.0, deprecated: 10.12)
```
into:
```
@available(OSX, introduced: 10.0, deprecated: 10.12)
// insert:    ^-----------^
```

Resolves [SR-4231](https://bugs.swift.org/browse/SR-4231).

### CODE REVIEW

**Comment by d-ronnqvist** [comment]
I've tried to keep the comments focused on why something is done, but I'm not sure if I ended up
with a good comment to code ratio or if it's too much.

File: test/Parse/diagnose_availability.swift [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+@available(OSX 10.6, *) // no error
+func availableSince10_6() {}
+
+@available(OSX, introduced: 10.0, deprecated: 10.12) // no error
+func introducedFollowedByDepreaction() {}
+
+@available(OSX 10.0, deprecated: 10.12)
+// expected-error@-1 {{'deprecated' can't be combined with shorthand specification 'OSX 10.0'}} {{15-15=, introduced:}}
+// expected-error@-2 {{expected declaration}}
```
Comment (CodaFi) [inline test/Parse/diagnose_availability.swift]: Is there a reasonable recovery path here?

File: lib/Parse/ParseStmt.cpp [hunk 2]
```diff
@@ -1103,7 +1103,47 @@ Parser::parseAvailabilitySpecList(SmallVectorImpl<AvailabilitySpec *> &Specs) {
       consumeToken();
       Status.setIsParseError();
     } else if (consumeIf(tok::comma)) {
-      // keep going.
```
Comment (CodaFi) [inline lib/Parse/ParseStmt.cpp]: In general, can you run git-clang-format on this patch.  There's like 4 or 5 places with some interesting indentation.

File: lib/Parse/ParseStmt.cpp [hunk 3]
```diff
@@ -1103,7 +1103,49 @@ Parser::parseAvailabilitySpecList(SmallVectorImpl<AvailabilitySpec *> &Specs) {
...
+      // There is more to parse in this list.
+
+      // Before continuing to parse the next specification, we check that it's
+      // also in the shorthand syntax and try to and provide a more specific
```
Comment (xwu) [inline lib/Parse/ParseStmt.cpp]: Drive-by nit: "and try to and provide"?

File: test/Parse/diagnose_availability.swift [hunk 4]
```diff
@@ -0,0 +1,27 @@
+// RUN: %target-typecheck-verify-swift 
+
+// SR-4231: Misleading/wrong error message for malformed @available
+
+@available(OSX 10.6, *) // no error
+func availableSince10_6() {}
+
+@available(OSX, introduced: 10.0, deprecated: 10.12) // no error
+func introducedFollowedByDepreaction() {}
```
Comment (xwu) [inline test/Parse/diagnose_availability.swift]: Another nit: here and below (not that it matters for your test), typo "Depreaction".

File: test/Parse/diagnose_availability.swift [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+// SR-4231: Misleading/wrong error message for malformed @available
+
+@available(OSX 10.6, *) // no error
+func availableSince10_6() {}
+
+@available(OSX, introduced: 10.0, deprecated: 10.12) // no error
+func introducedFollowedByDeprecated() {}
+
+@available(OSX 10.0, deprecated: 10.12)
```
Comment (CodaFi) [inline test/Parse/diagnose_availability.swift]: For visibility (got folded by the force-pushes): Can you make this recover?  There's a decl down there, but this second diagnostic makes it seems like there isn't.

Comment (d-ronnqvist) [inline test/Parse/diagnose_availability.swift]: I think that it probably could be made to recover, but I don't know how to do it. (I’m assuming that recovery means being able to emit the right attribute and continue parsing.)

As far as I understand how this works `parseAvailabilitySpecList` (this code) parses a comma separated list of short for availability specifications, and returns it to the `DAK_Available` case of `parseNewDeclAttribute` which turns each specification into their full attribute.

I think that in order to recover, it would need to pass some additional information back so that an `AvailableAttr` could be created with both an introduced and deprecated parameter, or so that the current token could be parsed differently. In either case, I need to consume tokens until the closing parenthesis.

Not that it would be a lot of work, but I don’t know if it would add much on top of the diagnostic and fix-it. That said; if you think it's worth doing, I would see it as a good learning opportunity.

Comment (CodaFi) [inline test/Parse/diagnose_availability.swift]: Recovery could be as simple as re-forming the attribute list as though the user had accepted your fixit, then just sending things on downstream.

**Comment by CodaFi** [comment]
@swift-ci please smoke tes

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

File: lib/Parse/ParseStmt.cpp [hunk 6]
```diff
@@ -1103,7 +1103,49 @@ Parser::parseAvailabilitySpecList(SmallVectorImpl<AvailabilitySpec *> &Specs) {
...
+              Text != "introduced") {
+            auto *PlatformSpec =
+                cast<PlatformVersionConstraintAvailabilitySpec>(Previous);
+
+            auto PlatformName = platformString(PlatformSpec->getPlatform());
+            auto PlatformNameEndLoc =
+                PlatformSpec->getPlatformLoc().getAdvancedLoc(
+                    PlatformName.size());
+
+            diag.fixItInsert(PlatformNameEndLoc, ", introduced:");
```
Comment (CodaFi) [inline lib/Parse/ParseStmt.cpp]: One last thing: Can you motivate this fixit insertion with either a separate note or a dependent clause in the error text that mentions your assumption here?

**Review by CodaFi — CHANGES_REQUESTED** [review]

**Comment by CodaFi** [comment]
Thanks @d-ronnqvist!  Don't worry yourself with the recovery thing for now.  I've got one last request then I think this can be put in.

File: lib/Parse/ParseStmt.cpp (hunk 6 above)
Comment (d-ronnqvist) [inline lib/Parse/ParseStmt.cpp]: Since the same error message is used both when there is and isn't a fix-it, I think it would be better to add a separate note and also move the fix-it to that note.

**Review by CodaFi — APPROVED** [review]

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
Thanks @d-ronnqvist!

⛵️
